modules/core/src/main/scala/sangria/execution/FutureResolver.scala

```diff
@@ -1534,6 +1535,124 @@ private[execution] class FutureResolver[Ctx](
     Result(errorReg, if (canceled) None else Some(marshaller.arrayNode(listBuilder.result())))
   }
 
+  private def trackDeprecation(ctx: Context[Ctx, _]): Unit = {
+    val fieldArgs = ctx.args
+    val visitedDirectives = mutable.Set[ast.Directive]()
+
+    def getArgValue(name: String, args: Args): Option[_] =
+      if (args.argDefinedInQuery(name)) {
+        if (args.optionalArgs.contains(name)) {
+          args.argOpt(name)
+        } else {
+          Some(args.arg(name))
+        }
+      } else {
+        None
+      }
+
+    def deprecatedArgsUsed(argDefs: List[Argument[_]], argValues: Args): List[Argument[_]] =
+      argDefs.filter { argDef =>
+        val argValue = getArgValue(argDef.name, argValues)
+        argDef.deprecationReason.isDefined && argValue.isDefined
+      }
+
+    def trackDeprecatedDirectiveArgs(astDirective: ast.Directive): Unit = {
+      // prevent infinite loop from directiveA -> arg -> directiveA -> arg ...
```

Comment:
I'm not sure if this is something we can actually run into with query parsing, but it is something that was possible to run into with tests